# Exploratory Data Analysis in Python

This code is main edited by Enze Wang

# Attention

1. Please put four .csv file in data folder and put this code in main folder instead of code folder. 

2. The four .csv file are from https://drive.google.com/drive/folders/16EfjTT4BNjj9D6pakG6MR0Bpt2Pj3Hfi

This code want to do a EDA for our data and find a good topic, for example, all pizza restaurant or all coffee business in our data. And we will output a good .csv for our teammates to deal with. 


In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import random


The most important .csv is business and review. 

In [2]:
business_city = pd.read_csv("data/business_city.csv")
review_city = pd.read_csv("data/review_city.csv")

/Users/enze/opt/anaconda3/envs/STAT628/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (6,16,30,41,44) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
business_city

Unnamed: 0             business_id                            name  \
0               0  pQeaRpvuhoEqudo3uymHIQ             The Empanadas House   
1               1  7Dv4_HAxsxvadEsT5fxQBg     Dependable Brakes & Exhaust   
2               2  M_guz7Dj7hX0evS672wIwA      Chocolate Shoppe Ice Cream   
3               3  03x6ZlJ7s39DHqfTJU1Ecg          Beechmont Country Club   
4               4  _4Oe9V-qTpU5iemM9bphlA    Goodyear Auto Service Center   
...           ...                     ...                             ...   
36322       36322  kCIvm-8hNfD-qrCxiWLsSQ                 The Smoke Shack   
36323       36323  AlwKuzyrrxlZ04ktebwmcA                       Pizza Pit   
36324       36324  TCkmgOfreqwSZMTlTGRQhg  Cinemark At Valley View and XD   
36325       36325  xVpE01l6ZXdEtVf5PkRpDg                           Julep   
36326       36326  BAVuLTDmpSzDCk37A5HjtQ               Bruegger's Bagels   

                             address           city state postal_code  \
0                     404 E Green St      Champaign    IL       61820   
1             1110 Saw Mill Run Blvd     Pittsburgh    PA       15226   
2                  2831 Parmenter St      Middleton    WI       53562   
3                 29600 Chagrin Blvd      Beachwood    OH       44122   
4      800 Fort Duquesne Blvd, Ste 1     Pittsburgh    PA       15222   
...                              ...            ...   ...         ...   
36322                 208 E Green St      Champaign    IL       61820   
36323                    13 Atlas Ct        Madison    WI       53714   
36324                  6001 Canal Rd    Valley View    OH       44125   
36325           829 E Washington Ave        Madison    WI       53703   
36326                27045 Lorain Rd  North Olmsted    OH       44070   

        latitude  longitude  stars  ...     Smoking  WheelchairAccessible  \
0      40.110446 -88.233073    4.5  ...         NaN                   NaN   
1      40.406667 -80.004450    5.0  ...         NaN                   NaN   
2      43.105310 -89.510142    3.5  ...         NaN                  True   
3      41.461448 -81.470127    4.5  ...         NaN                   NaN   
4      40.444614 -79.999344    3.5  ...         NaN                   NaN   
...          ...        ...    ...  ...         ...                   ...   
36322  40.110417 -88.236094    2.0  ...         NaN                   NaN   
36323  43.086793 -89.306868    2.5  ...         NaN                   NaN   
36324  41.399299 -81.623908    4.0  ...         NaN                  True   
36325  43.081022 -89.374006    4.0  ...  u'outdoor'                  True   
36326  41.416568 -81.921396    3.0  ...         NaN                   NaN   

          WiFi       Friday       Monday   Saturday     Sunday     Thursday  \
0          NaN  11:30-14:30  11:30-14:30        NaN        NaN  11:30-14:30   
1          NaN     8:0-17:0     8:0-17:0   8:0-17:0        NaN     8:0-17:0   
2        u'no'    11:0-22:0      0:0-0:0  11:0-22:0  11:0-21:0    11:0-21:0   
3          NaN          NaN          NaN        NaN        NaN          NaN   
4          NaN     7:0-18:0     7:0-18:0        NaN        NaN     7:0-18:0   
...        ...          ...          ...        ...        ...          ...   
36322      NaN          NaN          NaN        NaN        NaN          NaN   
36323    u'no'     11:0-0:0     11:0-0:0   11:0-0:0   11:0-0:0     11:0-0:0   
36324      NaN          NaN          NaN        NaN        NaN          NaN   
36325  u'free'    16:0-23:0    16:0-22:0  16:0-23:0  10:0-14:0    16:0-22:0   
36326      NaN    5:30-15:0    5:30-15:0   6:0-15:0   6:0-15:0    5:30-15:0   

           Tuesday    Wednesday  
0      11:30-14:30  11:30-14:30  
1         8:0-17:0          NaN  
2        11:0-21:0    11:0-21:0  
3              NaN          NaN  
4         7:0-18:0     8:0-16:0  
...            ...          ...  
36322          NaN          NaN  
36323     11:0-0:0     11:0-0:0  
36324          NaN       

In [4]:
review_city

Unnamed: 0               review_id                 user_id  \
0                0  LG2ZaYiOgpr2DK_90pYjNw  V34qejxNsCbcgD8C0HVk-Q   
1                1  qWHp2l2lysENZObh6MAulA  5lb0POg2t-AkMFx6603cYQ   
2                2  Ne_2CSfcKIqXHmv_KC0uqg  6sJN_HlM_uwpfLJ1puf1Fg   
3                3  2mLlQ9n5Ln_fykOuX_b0-A  EZ6g9a17UdMjtR-wrZebmA   
4                4  Ymb8ZYPg2e8dvqhdj3Gkqw  hHFSpKcnDtJOs6M-JKFZ8Q   
...            ...                     ...                     ...   
942022      942022  YWfpUgsEKYg_J_dx3O-RPQ  XYCyePb6v-RKdm_THQQQbQ   
942023      942023  dVCtr4nBevsAfUzVqBn8GQ  kqytC4gk1EPGq6LoIYIoiA   
942024      942024  0_skOI7rYQwlVn0hfeIQTA  RfYEn-JHiP7QVovc-3ltgA   
942025      942025  b2nh40wvzDHry1A_9tT_9w  pjj4Y76C7jHNpH4WOG6slQ   
942026      942026  kBdIrB1J41TDQArBWOpN0g  VwVHW0hunS1NXceJJtfVrA   

                   business_id  stars  useful  funny  cool  \
0       HQl28KMwrEKHqhFrrDqVNQ      5       1      0     0   
1       ZlCSsWS07JulSBIQlrbF5w      5       1      0     0   
2       oVuZtlCFg_zF090Nhtp0Ow      4       3      0     1   
3       wSAGILK2CFbq5BVZ5D67-A      3       0      0     0   
4       oCvxJBOs-Mbg-4oanPO7lQ      4       0      0     0   
...                        ...    ...     ...    ...   ...   
942022  hWzbOi1-eS8PmPyZRV8Lew      5       0      0     0   
942023  NC4_GCUAB9fuvhWgl8bu5Q      5       0      0     0   
942024  0hlQbEOdqNBagHP6xFgRRQ      5       0      0     0   
942025  QayOwK3LVRVniGnT6Gzveg      1       0      0     0   
942026  zxg3vChBI__R6q7ACtyokw      1       1      0     0   

                                                     text                 date  
0       I love Deagan's. I do. I really do. The atmosp...  2015-12-05 03:18:11  
1       Called for a 5:15 morning reservation. The dri...  2015-05-26 10:36:47  
2       A close friend was in town and so instead of t...  2018-02-01 19:15:00  
3       The service is attentive. The decor attractive...  2015-06-21 00:07:59  
4       The ingredients are always fresh and I like th...  2016-10-26 14:53:42  
...                                                   ...                  ...  
942022  AMAZING lattes and scones.  Nice relaxing atmo...  2019-12-02 17:24:13  
942023  The staff is very friendly and helpful. The fo...  2019-12-02 20:13:56  
942024  This salon is very clean, professional and upb...  2019-12-09 21:48:39  
942025  Went after work to order dinner. I asked the c...  2019-12-10 00:39:24  
942026  I should have read the pattern of complaints o...  2019-06-15 03:00:07  

[942027 rows x 10 columns]

# Null value

In [5]:
pd.set_option("display.max_rows", 200)
pd.set_option("display.max_columns", 100)
pd.set_option("display.max_colwidth", 200)
display(pd.DataFrame(business_city.isnull().sum()))

0
Unnamed: 0                      0
business_id                     0
name                            0
address                       776
city                            0
state                           0
postal_code                    42
latitude                        0
longitude                       0
stars                           0
review_count                    0
is_open                         0
attributes                   4678
categories                     95
hours                        7859
AcceptsInsurance            35537
AgesAllowed                 36315
Alcohol                     26218
Ambience                    25788
BYOB                        36182
BYOBCorkage                 36012
BestNights                  35038
BikeParking                 19893
BusinessAcceptsBitcoin      33073
BusinessAcceptsCreditCards   7778
BusinessParking             14771
ByAppointmentOnly           28058
Caters                      27444
CoatCheck                   35179
Corkage                     36132
DietaryRestrictions         36316
DogsAllowed                 33615
DriveThru                   35650
GoodForDancing              35188
GoodForKids                 23200
GoodForMeal                 29640
HairSpecializesIn           36169
HappyHour                   33051
HasTV                       25791
Music                       34362
NoiseLevel                  27363
Open24Hours                 36325
OutdoorSeating              24960
RestaurantsAttire           26510
RestaurantsCounterService   36326
RestaurantsDelivery         24687
RestaurantsGoodForGroups    25322
RestaurantsPriceRange2      15306
RestaurantsReservations     25255
RestaurantsTableService     32657
RestaurantsTakeOut          23006
Smoking                     35418
WheelchairAccessible        31600
WiFi                        24611
Friday                       8284
Monday                      10542
Saturday                    11011
Sunday                      18439
Thursday                     8252
Tuesday                      8756
Wednesday                    8440

There is about 942027 business, and there is no null value in latitude, longitude, stars, review_count, is_open, city, state, Unnamed, business_id, name. 

# Distance Clean

In [6]:
set(business_city["city"])

{'ARSENAL',
 'Akron',
 'Allegheny',
 'Allentown',
 'Allison Park',
 'Ambridge',
 'Amherst',
 'Arnold',
 'Aspinwall',
 'Auburn',
 'Auburn Twp',
 'Aurora',
 'Austinburg',
 'Avalon',
 'Avon',
 'Avon Lake',
 'Avon lake',
 'Bainbridge',
 'Bainbridge Township',
 'Bakerstown',
 'Baldwin',
 'Banksville',
 'Barberton',
 'Bath',
 'Bay Village',
 'Beachwood',
 'Beachwood ',
 'Bedford',
 'Bedford Heights',
 'Bedford Hts',
 'Bedford Hts.',
 'Belle Vernon',
 'Belleville',
 'Bellevue',
 'Bellvue',
 'Ben Avon',
 'Bentleyville',
 'Berea',
 'Berry',
 'Bethel Park',
 'Black Earth',
 'Blawnox',
 'Bloomfield',
 'Blue Mounds',
 'Boston',
 'Boston Heights',
 'Braddock',
 'Bradford Woods',
 'Bradfordwoods',
 'Bratenahl',
 'Brecksville',
 'Brentwood',
 'Bridgeview',
 'Bridgeville',
 'Broadlands',
 'Broadview Heights',
 'Broadview Hts',
 'Brook Park',
 'Brookline',
 'Brooklyn',
 'Brooklyn Heights',
 'Brooklyn Hts.',
 'Brookpark',
 'Brunswick',
 'Brunswick Hills',
 'Buckeye - Shaker',
 'Buena Vista',
 'Burton',


In [7]:
set(business_city["state"])

{'IL', 'OH', 'PA', 'WI'}

As professor mentioned in data description, all data are from Madison (U.S.), Cleveland (U.S.), Pittsburgh (U.S.) and Urbana-Champaign (U.S.). Therefore, all business are from four states, Madison, WI, Cleveland, OH, Pittsburgh, PA and Urbana-Champaign, IL. All city in city column are neighbor around them. For example Allison Park are around Pittsburgh.

Another way to check whether all business are from the four cities is to use latitude and longitude. 

We check the latitude and longitude of Madison, Cleveland, Pittsburgh and Urbana-Chamaign and find the min distance between one business with four of them. 

In [8]:
Latitude=business_city["latitude"]
Longitude=business_city["longitude"]

Maidson_Latitude=43.0731635
Maidson_Longitude=-89.4003412

Cleveland_Latitude=41.4990361 
Cleveland_Longitude=-81.6927877

Pittsburgh_Latitude=40.4356121
Pittsburgh_Longitude=-79.9928454

Urbana_Champaign_Latitude=40.1124904
Urbana_Champaign_Longitude=-88.2259212



In [9]:
from math import sin, cos, sqrt, atan2, radians
def Distance(Lat1, Lon1, Lat2, Lon2):
    Lat1=radians(Lat1)
    Lat2=radians(Lat2)
    Lon1=radians(Lon1)
    Lon2=radians(Lon2)
    DLat=Lat1-Lat2
    DLon=Lon1-Lon2
    a=sin(DLat/2)**2+cos(Lat1)*cos(Lat2)*sin(DLon/2)**2
    c=2*atan2(sqrt(a),sqrt(1-a))
    return 6373*c

In [10]:
Distance(Maidson_Latitude, Maidson_Longitude, Cleveland_Latitude, Cleveland_Longitude)

657.6816098580905

The real distance from Madison to Cleveland is 657.400931km, so there is no problems of our Distance function. 

In [12]:
r,c=business_city.shape
r

36327

In [13]:
distance=[]
for i in range(r):
    Goal_Lat=float(Latitude[i])
    Goal_Lon=float(Longitude[i])
    Dis_Mad=Distance(Goal_Lat,Goal_Lon,Maidson_Latitude,Maidson_Longitude)
    Dis_Cle=Distance(Goal_Lat,Goal_Lon,Cleveland_Latitude,Cleveland_Longitude)
    Dis_Pit=Distance(Goal_Lat,Goal_Lon,Pittsburgh_Latitude,Pittsburgh_Longitude)
    Dis_Urb=Distance(Goal_Lat,Goal_Lon,Urbana_Champaign_Latitude,Urbana_Champaign_Longitude)
    Min_Dis=min(Dis_Mad,Dis_Cle,Dis_Pit,Dis_Urb)
    distance.append([i,Min_Dis])

In [14]:
distance

[[0, 0.6494759695259577],
 [1, 3.3661462996773537],
 [2, 9.609157965186366],
 [3, 19.019883464524636],
 [4, 1.1424843074474296],
 [5, 2.854768818003282],
 [6, 15.563757223271915],
 [7, 1.8633417020791454],
 [8, 19.058136891309182],
 [9, 6.790898921249643],
 [10, 25.569881396029146],
 [11, 2.1015290925833554],
 [12, 2.9563753312695815],
 [13, 12.095425923928106],
 [14, 37.68999349217335],
 [15, 15.33682043229218],
 [16, 15.969807247867031],
 [17, 14.083811971728856],
 [18, 9.136012285734683],
 [19, 7.938232139485963],
 [20, 16.31670581544371],
 [21, 23.403030707029902],
 [22, 11.091357058889377],
 [23, 4.087328185040756],
 [24, 17.327763643505186],
 [25, 8.915908096323228],
 [26, 7.492259921025159],
 [27, 36.11006514169806],
 [28, 40.108205351953586],
 [29, 14.296191272376493],
 [30, 40.935580103472624],
 [31, 1.557770377283669],
 [32, 9.317696039653153],
 [33, 12.575506998987608],
 [34, 3.602619756636366],
 [35, 6.330629265504701],
 [36, 22.600635692957564],
 [37, 16.09295140798034],
 

In [15]:
sorted(distance,key=lambda l:l[1], reverse=True)

[[33040, 2329.645313019225],
 [10833, 2240.7430697274112],
 [22678, 2238.8819902699493],
 [6872, 2228.509835950993],
 [22776, 595.6480877769187],
 [34559, 584.6871629486197],
 [18116, 303.221123406281],
 [24041, 296.8225577323126],
 [31467, 270.6046821271112],
 [21037, 183.3763871543956],
 [19149, 153.3506366427876],
 [6831, 54.870884144703986],
 [26692, 53.802418702864536],
 [14287, 53.77443932347636],
 [7656, 53.677291239443655],
 [28370, 53.24862306283845],
 [12910, 53.11269204725765],
 [2761, 52.94783902631065],
 [24048, 52.83628986252499],
 [29876, 52.799062047006146],
 [7123, 52.77361210089823],
 [24961, 52.75521675319252],
 [23393, 52.639613589587526],
 [24963, 52.251674666265735],
 [28820, 52.24347111958175],
 [8671, 52.239218217775466],
 [32412, 52.227584167448484],
 [21047, 52.19806256445451],
 [21122, 52.04690552701736],
 [9827, 51.961430672395224],
 [20279, 51.96037870557648],
 [22160, 51.94933403755965],
 [13887, 51.89642026975997],
 [32135, 51.86664783133546],
 [29886, 51

As you can see that, there are some business have min 2400 km from the four cities, for example, the ID 19149:
    

In [16]:
business_city.iloc[21037,[7,8]]

latitude     41.9769
longitude   -87.7295
Name: 21037, dtype: object

It is in Chicago. And ID 33040

In [17]:
business_city.iloc[33040,[7,8]]

latitude     36.2915
longitude   -115.286
Name: 33040, dtype: object

It is in Las Vegas, so we decide to remove distance larger than 55km. 

In [18]:
business_city_clean1=business_city.drop([33040,10833,22678,6872,22776,34559,18116,24041,31467,21037,19149])
r,c=business_city_clean1.shape
business_city_clean1

Unnamed: 0             business_id                            name  \
0               0  pQeaRpvuhoEqudo3uymHIQ             The Empanadas House   
1               1  7Dv4_HAxsxvadEsT5fxQBg     Dependable Brakes & Exhaust   
2               2  M_guz7Dj7hX0evS672wIwA      Chocolate Shoppe Ice Cream   
3               3  03x6ZlJ7s39DHqfTJU1Ecg          Beechmont Country Club   
4               4  _4Oe9V-qTpU5iemM9bphlA    Goodyear Auto Service Center   
...           ...                     ...                             ...   
36322       36322  kCIvm-8hNfD-qrCxiWLsSQ                 The Smoke Shack   
36323       36323  AlwKuzyrrxlZ04ktebwmcA                       Pizza Pit   
36324       36324  TCkmgOfreqwSZMTlTGRQhg  Cinemark At Valley View and XD   
36325       36325  xVpE01l6ZXdEtVf5PkRpDg                           Julep   
36326       36326  BAVuLTDmpSzDCk37A5HjtQ               Bruegger's Bagels   

                             address           city state postal_code  \
0                     404 E Green St      Champaign    IL       61820   
1             1110 Saw Mill Run Blvd     Pittsburgh    PA       15226   
2                  2831 Parmenter St      Middleton    WI       53562   
3                 29600 Chagrin Blvd      Beachwood    OH       44122   
4      800 Fort Duquesne Blvd, Ste 1     Pittsburgh    PA       15222   
...                              ...            ...   ...         ...   
36322                 208 E Green St      Champaign    IL       61820   
36323                    13 Atlas Ct        Madison    WI       53714   
36324                  6001 Canal Rd    Valley View    OH       44125   
36325           829 E Washington Ave        Madison    WI       53703   
36326                27045 Lorain Rd  North Olmsted    OH       44070   

        latitude  longitude  stars  review_count  is_open  \
0      40.110446 -88.233073    4.5             5        1   
1      40.406667 -80.004450    5.0             5        1   
2      43.105310 -89.510142    3.5             6        1   
3      41.461448 -81.470127    4.5             7        1   
4      40.444614 -79.999344    3.5            12        1   
...          ...        ...    ...           ...      ...   
36322  40.110417 -88.236094    2.0             5        0   
36323  43.086793 -89.306868    2.5            19        1   
36324  41.399299 -81.623908    4.0           120        1   
36325  43.081022 -89.374006    4.0            95        0   
36326  41.416568 -81.921396    3.0            13        0   

                                                                                                                                                                                                    attributes  \
0      {'RestaurantsAttire': "u'casual'", 'RestaurantsTakeOut': 'True', 'BusinessAcceptsCreditCards': 'True', 'NoiseLevel': "u'quiet'", 'GoodForKids': 'False', 'RestaurantsReservations': 'False', 'Restau...   
1                                                                                                                                                                       {'BusinessAcceptsCreditCards': 'True'}   
2      {'BikeParking': 'True', 'Caters': 'True', 'WheelchairAccessible': 'True', 'RestaurantsPriceRange2': '1', 'BusinessParking': "{'garage': False, 'street': True, 'validated': False, 'lot': True, 'val...   
3                                                                                                                                               {'GoodForKids': 'True', 'BusinessAcceptsCreditCards': 'False'}   
4                                                                                                                                                                       {'BusinessAcceptsCreditCards': 'True'}   
...                                                                                                                                                                                                        ...   
36

In [19]:
Latitude=business_city_clean1["latitude"].tolist()
Longitude=business_city_clean1["longitude"].tolist()



In [20]:
new_distance=[]
for i in range(r):
    Goal_Lat=float(Latitude[i])
    Goal_Lon=float(Longitude[i])
    Dis_Mad=Distance(Goal_Lat,Goal_Lon,Maidson_Latitude,Maidson_Longitude)
    Dis_Cle=Distance(Goal_Lat,Goal_Lon,Cleveland_Latitude,Cleveland_Longitude)
    Dis_Pit=Distance(Goal_Lat,Goal_Lon,Pittsburgh_Latitude,Pittsburgh_Longitude)
    Dis_Urb=Distance(Goal_Lat,Goal_Lon,Urbana_Champaign_Latitude,Urbana_Champaign_Longitude)
    Min_Dis=min(Dis_Mad,Dis_Cle,Dis_Pit,Dis_Urb)
    new_distance.append([i,Min_Dis])

In [21]:
sorted(new_distance,key=lambda l:l[1], reverse=True)

[[6831, 54.870884144703986],
 [26684, 53.802418702864536],
 [14285, 53.77443932347636],
 [7655, 53.677291239443655],
 [28362, 53.24862306283845],
 [12908, 53.11269204725765],
 [2761, 52.94783902631065],
 [24040, 52.83628986252499],
 [29868, 52.799062047006146],
 [7122, 52.77361210089823],
 [24953, 52.75521675319252],
 [23386, 52.639613589587526],
 [24955, 52.251674666265735],
 [28812, 52.24347111958175],
 [8670, 52.239218217775466],
 [32403, 52.227584167448484],
 [21042, 52.19806256445451],
 [21117, 52.04690552701736],
 [9826, 51.961430672395224],
 [20275, 51.96037870557648],
 [22155, 51.94933403755965],
 [13885, 51.89642026975997],
 [32126, 51.86664783133546],
 [29878, 51.860773349791536],
 [23520, 51.853990424799775],
 [1708, 51.84960240549111],
 [30661, 51.833200527089936],
 [26963, 51.83078721149567],
 [28166, 51.81818199336575],
 [16159, 51.80876680671499],
 [25603, 51.780374868315086],
 [32496, 51.75784147567597],
 [15296, 51.75282670748638],
 [33739, 51.75282670748638],
 [10635,

Now there is no problem on distance. 

# Category

No matter we choose pizza or coffee or other buiness, there must be:
    
1. Enough business for us to study.

2. Enough review for us to study.

Therefore we first count categories from our table.

In [56]:
set(business_city_clean1["categories"])

{nan,
 'Restaurants, Bars, American (New), Cocktail Bars, Pizza, Nightlife',
 'Property Management, Home Services, Real Estate, Real Estate Services, Apartments, Real Estate Agents',
 'Food, Restaurants, Breweries, American (New), Nightlife, Event Planning & Services, Pubs, Caterers, Bars',
 'Fire Departments, Public Services & Government',
 'Food, Ice Cream & Frozen Yogurt, Vegan, Restaurants',
 'Event Planning & Services, American (New), Music Venues, Bars, Restaurants, American (Traditional), Venues & Event Spaces, Nightlife, Musicians, Arts & Entertainment, Performing Arts',
 'Beauty & Spas, Hair Removal',
 'Bars, American (Traditional), Pizza, Nightlife, Restaurants',
 'Coffee & Tea, Food, Juice Bars & Smoothies, Desserts, Restaurants, Breakfast & Brunch, Sandwiches',
 'Nurseries & Gardening, Shopping, Home & Garden, Home Services, Tree Services, Landscaping, Lawn Services',
 'Car Dealers, Used Car Dealers, Automotive, Auto Repair',
 'Auto Repair, Automotive, Oil Change Stations, 

In [57]:
All_Category=[]
for category in business_city_clean1["categories"]:
    if isinstance(category, str)==True: 
        All_Category.extend(category.split(','))

In [58]:
Unique_Category=set(All_Category)
Category_Count=[]
for category in Unique_Category:
    Category_Count.append([category,All_Category.count(category)])

In [59]:
sorted(Category_Count,key=lambda l:l[1], reverse=True)

[[' Restaurants', 8988],
 [' Food', 4860],
 [' Shopping', 4705],
 ['Restaurants', 3676],
 [' Nightlife', 2718],
 [' Bars', 2513],
 [' Home Services', 2269],
 [' Beauty & Spas', 2161],
 [' Automotive', 1926],
 [' Local Services', 1905],
 [' Event Planning & Services', 1785],
 [' American (Traditional)', 1779],
 [' Health & Medical', 1628],
 ['Food', 1624],
 ['Shopping', 1524],
 [' Sandwiches', 1522],
 [' Pizza', 1360],
 [' Active Life', 1356],
 [' Fast Food', 1263],
 [' American (New)', 1252],
 [' Home & Garden', 1201],
 [' Fashion', 1161],
 [' Burgers', 1142],
 [' Auto Repair', 1071],
 [' Coffee & Tea', 1060],
 ['Beauty & Spas', 1013],
 [' Arts & Entertainment', 929],
 [' Hotels & Travel', 922],
 [' Breakfast & Brunch', 917],
 ['Automotive', 872],
 [' Hair Salons', 868],
 ['Home Services', 837],
 [' Italian', 811],
 [' Specialty Food', 797],
 [' Nail Salons', 704],
 ['Pizza', 675],
 ['Health & Medical', 663],
 [' Fitness & Instruction', 651],
 [' Wine & Spirits', 646],
 [' Salad', 639]

Table above count category in our business_city.csv. As you can see, Restaurants occurs the most time. Pizza and Coffee & Tea occurs about 1000 times in our data. 

# Pizza Business

There are too many categories in our data, because many business has multiple categories. We focus on pizza business first

In [60]:
business_city_clean1["pizza"]=business_city_clean1["categories"].str.contains('Pizza|pizza')
business_city_clean1.pizza.value_counts()

False    34186
True      2035
Name: pizza, dtype: int64

There are about 2000 pizza business. We select pizza business.

In [61]:
business_city_pizza=business_city_clean1.loc[business_city_clean1["pizza"]==True]

In [62]:
business_city_pizza

Unnamed: 0             business_id                        name  \
8               8  ZkzutF0P_u0C0yTulwaHkA            Lelulos Pizzeria   
16             16  39lLJK_rrYY2NYomSsQdUA               Marco's Pizza   
17             17  0y6alZmSLnPzmG5_kP5Quw                 J J's Pizza   
53             53  _Kp1IPTi17wBywYOd30raA                   Pizza Hut   
54             54  1YC7AbQMlNb5mLUlwwaa3w             Bites and Brews   
...           ...                     ...                         ...   
36270       36270  Qtv6iGDAHBmAKBgl40JgLw     Blaze Fast-Fire'd Pizza   
36271       36271  ZZGs9idRiKRHue2RyXw9uA               Marco's Pizza   
36296       36296  v8GgrqLTs_NZZl2tkyvOgA  Cleveland Clinic Cafeteria   
36309       36309  qWcaGio9JsqrvgmQiIRtIQ         Country Style Pizza   
36323       36323  AlwKuzyrrxlZ04ktebwmcA                   Pizza Pit   

                             address           city state postal_code  \
8                311 Unity Center Rd           Plum    PA       15239   
16                24335 Chagrin Blvd      Beachwood    OH       44122   
17                   20542 Lorain Rd  Fairview Park    OH       44126   
53              8609 University Blvd     Coraopolis    PA       15108   
54                5750 Ellsworth Ave     Pittsburgh    PA       15232   
...                              ...            ...   ...         ...   
36270            2610 N Prospect Ave      Champaign    IL       61822   
36271              5877 Broadview Rd          Parma    OH       44134   
36296  9500 Euclid Ave, Bldg C, Fl 1      Cleveland    OH       44195   
36309                  536 Dewey Ave    Bridgeville    PA       15017   
36323                    13 Atlas Ct        Madison    WI       53714   

        latitude  longitude  stars  review_count  is_open  \
8      40.489996 -79.779288    4.0            31        1   
16     41.465789 -81.506349    2.5            11        1   
17     41.448341 -81.847644    4.5            21        1   
53     40.507268 -80.222632    2.0            14        1   
54     40.455696 -79.930883    3.5            85        0   
...          ...        ...    ...           ...      ...   
36270  40.147209 -88.254940    3.5            55        1   
36271  41.403398 -81.689779    3.0            10        1   
36296  41.501753 -81.621448    3.0            12        1   
36309  40.356300 -80.110600    4.0            39        1   
36323  43.086793 -89.306868    2.5            19        1   

                                                                                                                                                                                                    attributes  \
8      {'RestaurantsPriceRange2': '1', 'BusinessAcceptsCreditCards': 'True', 'GoodForKids': 'True', 'OutdoorSeating': 'False', 'HasTV': 'True', 'WiFi': "'no'", 'BusinessParking': "{'garage': False, 'stre...   
16     {'RestaurantsDelivery': 'True', 'GoodForKids': 'True', 'RestaurantsReservations': 'False', 'BusinessAcceptsCreditCards': 'True', 'RestaurantsAttire': "u'casual'", 'NoiseLevel': "u'quiet'", 'GoodFo...   
17     {'NoiseLevel': "u'quiet'", 'WiFi': "'no'", 'BusinessParking': "{'garage': False, 'street': False, 'validated': False, 'lot': False, 'valet': False}", 'OutdoorSeating': 'False', 'RestaurantsReserva...   
53     {'NoiseLevel': "u'quiet'", 'RestaurantsGoodForGroups': 'True', 'Caters': 'False', 'HasTV': 'True', 'WiFi': "u'no'", 'BusinessAcceptsCreditCards': 'True', 'RestaurantsAttire': "u'casual'", 'Restaur...   
54     {'GoodForKids': 'False', 'BusinessAcceptsCreditCards': 'True', 'BusinessParking': "{'garage': False, 'street': True, 'validated': False, 'lot': False, 'valet': False}", 'OutdoorSeating': 'True', '...   
...                                                                                                                                                                                                        ...   
36270  {'DriveThru': 'False', 'RestaurantsTakeOut'

In [63]:
set(business_city_pizza["categories"])

{'Active Life, Mini Golf, Ice Cream & Frozen Yogurt, Pizza, Restaurants, Food',
 'American (New), Bars, Sandwiches, Restaurants, Sports Bars, Nightlife, Pizza, Wraps',
 'American (New), Burgers, Nightlife, Restaurants, Bars, Sandwiches, Pizza, Sports Bars',
 'American (New), Burgers, Pizza, Restaurants',
 'American (New), Chicken Wings, Nightlife, Pubs, Bars, American (Traditional), Restaurants, Sandwiches, Sports Bars, Pizza',
 'American (New), Gluten-Free, Salad, Italian, American (Traditional), Pizza, Restaurants',
 'American (New), Mediterranean, Restaurants, Vegetarian, Pizza, Desserts, Food, Vegan, Middle Eastern',
 'American (New), Middle Eastern, Ethnic Food, Food, Pizza, Specialty Food, Greek, Restaurants',
 'American (New), Nightlife, Sandwiches, Restaurants, Cocktail Bars, Wine Bars, Bars, Pizza',
 'American (New), Nightlife, Seafood, Pizza, Bars, Restaurants, Wine Bars',
 'American (New), Pizza, Mexican, Restaurants',
 'American (New), Pizza, Restaurants',
 'American (New),

In [64]:
pizza_id=business_city_pizza["business_id"].tolist()
pizza_id

['ZkzutF0P_u0C0yTulwaHkA',
 '39lLJK_rrYY2NYomSsQdUA',
 '0y6alZmSLnPzmG5_kP5Quw',
 '_Kp1IPTi17wBywYOd30raA',
 '1YC7AbQMlNb5mLUlwwaa3w',
 'crBw_Hq3gCkppirgjnIl4A',
 'cPNolx5Ae2znDZNA7GG5BQ',
 '7LzhLKAwS7PHUvaPIW833A',
 'ijNQXC2cBhbkf6iB6D8OJw',
 'LTPcaw-_IqOfTSDxe8811A',
 '7-OCdlOQQCHAqYDI_elK6Q',
 '00J65e-ewht6GZ9fXZdP1w',
 'isao1F5FW5kvQl402UlJ3g',
 'fhTTP5HRqOzmFw-31Qe7QA',
 'x1Cd2gSPZVeNxDPzzqlClQ',
 '4Vig7oCo9P-j2OnPUfaXCA',
 'sytLExNDyQuELMDguVUqhA',
 'mSWNf1muSy9wgGtMRgt7Fg',
 '_3-gItHIdBvQ9XrDrIQLXg',
 '2PBPh8LTwt_mXP0GUYa9NQ',
 'wTBfpTjdWG_zoE62B_bLtA',
 'EJGv2la6CcZmEIpkE4hzyg',
 'nsw3tlB-hwz4KbPYcuaSww',
 'goMccfETQlUOb9_USl4BUg',
 'ZfCzpZNYSWeA3Ouku2eqlA',
 'h8MMgum-XUQhU4xWCyQWLQ',
 'a3ZqjElMgAe924sbXZVcvA',
 'BQMVcJQV70gUQjwi7-MrSg',
 'y2lttTrB3kyoEG6kDSONbg',
 'GqH0U0E3Te5pUUjyiIzNBQ',
 'uE6hI5_i4QVq12xU99xtqA',
 'Qy_FDRtL-f-syUlEKpg_6A',
 'yU7GLvB07AEGhsoCc9scQg',
 'TQqhfjlPVEz41K_wAogQSQ',
 'vZmKNy61z53RnxBTCy86YQ',
 'gbDZhohLAOL1ufdVWHMTxQ',
 'Zp2vJfHciG9IiXfmtH5Wdw',
 

In [65]:
review_city_pizza=review_city[review_city["business_id"].isin(pizza_id)]

In [66]:
review_city_pizza

Unnamed: 0               review_id                 user_id  \
26              26  hfXVxrlD6fr1kXfz6sR13Q  7nIzgNF7YzI-UAyYr3y9hw   
28              28  EDQpLLFiAxeSxWI5ipwzBQ  SkueY7hVsyXmYItELoodGw   
43              43  RWmT90rP7wYo57dG1cyvww  n3ahpmnEXsZYj3nbVMoUmg   
57              57  0jduztItUhOh3FzKGjMxZQ  lVbPEaNUxFjwUydjentrYw   
59              59  sxfmzhMvZ8LlI1x4fhNjKg  SQKRGzDa7qd8qHb13Z2-Hg   
...            ...                     ...                     ...   
941979      941979  gcNRaJMDlSVs8YlqwpsRuw  _Rgn4SZT2wXyQnW9AD9Tog   
941995      941995  yFxDOdiDb8HPaK1io81kiw  4r33dXcE1oYZxjONrhxTiA   
942002      942002  Z_xmoLgIMB4As90t0a5eBQ  Iy6HkAzFcpeU7GF0e1FRSA   
942017      942017  CzYRFCnkbqOVb3WIMMrbfw  TP9Tz7gcAfK-BeT16qpZ2w   
942018      942018  ZhUeFHPAjUbEKRnYO_pl_w  WS-GsWbN4ltzigWl01t8Xw   

                   business_id  stars  useful  funny  cool  \
26      J9f-9Prw2YVM-fiZqv2fmQ      2       0      1     1   
28      lpel3pRFAzlow7u2TJMz4Q      4       3      0     0   
43      Zja7ykq3icDzkN20XX6j1w      5       1      0     1   
57      hr0vxeTzlHx6fTpSSmu2AA      5       1      0     0   
59      F4kjKxXXyn3e-BYVb1mWcg      4       0      0     0   
...                        ...    ...     ...    ...   ...   
941979  nR3u-wQ99bop6IdGoU38KQ      1       1      0     0   
941995  yT1Ad3auuw1jAJEZHr7Yqw      5       1      1     1   
942002  R13VkcyUIU3xG2fRFbBvBA      4       3      3     3   
942017  M-caugKzJ9qLOmCBfD9Nyg      1       1      0     0   
942018  o_dWM_hlpgjjd7F15UPjZw      3       0      0     0   

                                                                                                                                                                                                           text  \
26      Parking is tough to get here so I don't get to go much. I also get a little cautious sometimes as they frequently have multiple questionable looking characters.\n\nPro's:\n- Legendary fries that u...   
28      The server had an extensive knowledge of wines & their various menu selections and it was clear that we were her top priority. The pizza and dessert were amazing. My only complaint is that the cha...   
43      Fast Forward\n\nEvery trip from NYC to Pitt includes a visit to Aiello's. And if we know anyone driving East, we request a transfer!\n\nIf you are in search of fancy "gourmet" pizza, go elsewhere....   
57      By far my favorite Cleveland/Akron restaurant!  Great food, very consistent.  Bartenders are always great.   If it is on the menu... Try it.   Only the Navajo tacos have been so so.  Everything el...   
59      All food is good here I usually get the Italian sub or pizza. Would recommend. Very busy on Friday and Saturday. If you can pick it up usually takes about an hour for delivery. And if you order si...   
...                                                                                                                                                                                                         ...   
941979  I'm originally from the other side of Ohio and I've always really enjoyed jets pizza. However, the people that run the store just absolutely don't know what they're doing. Orders are wrong so ofte...   
941995  Yes, yes, YES!! We loved going to the place in Shadyside, and were ecstatic when we learned that they were opening this place up in Fox Chapel. My wife and I have been enjoying their pizza for qui...   
942002  Went here for a lunch meeting amidst s typical suburban strip mall, and I was pleasantly surprised by the  decor as I walked in. \n\nWe sat at the bar and was greeted by the friendly bartender, wh...   
942017  My wife and I have eaten at La Brioche at least a dozen times.  The food is nearly always excellent, but the service is uneven.  We went in for Sunday brunch and found the restaurant maybe half-fu...   
942018  First let me say Lindys in Columbus is one of my favorite restaurants, so I was anxious to tr

There are about 76000 reviews about pizza business. 

# Coffee Business

In [67]:
business_city_clean1["coffee"]=business_city_clean1["categories"].str.contains('Coffee|Cafe|coffee|cafe|cafes|Cafes')
business_city_clean1.coffee.value_counts()

False    34486
True      1735
Name: coffee, dtype: int64

In [68]:
business_city_coffee=business_city_clean1.loc[business_city_clean1["coffee"]==True]

In [69]:
set(business_city_coffee["categories"])

{'Restaurants, Sandwiches, Breakfast & Brunch, Food, Coffee & Tea',
 'Restaurants, Cafeteria',
 'Food, Restaurants, Sandwiches, Salad, Coffee & Tea, Cafes',
 'Restaurants, American (New), Cocktail Bars, Cafes, Comfort Food, Nightlife, Bars',
 'Donuts, Coffee & Tea, Food, Bakeries',
 'Automotive, Coffee & Tea, Gas Stations, Food, Convenience Stores',
 'Cafes, Beauty & Spas, Fashion, Shopping, Restaurants, Tattoo, Arts & Entertainment, Art Galleries, Food, Coffee & Tea, Piercing',
 'Food, Restaurants, Coffee & Tea, Cafes',
 'Restaurants, Coffee & Tea, Food, Fast Food, Burgers',
 'Restaurants, Burgers, Fast Food, Coffee & Tea, Food',
 'Diners, Restaurants, Cafes, Food, Breakfast & Brunch, Sandwiches, Food Delivery Services, Caterers, Event Planning & Services',
 "Shopping, Fashion, Cafes, Women's Clothing, Restaurants",
 'American (Traditional), Breakfast & Brunch, Cafes, Restaurants, Food, Desserts, Bakeries',
 'Coffee & Tea, Restaurants, Food, Breakfast & Brunch, American (Traditional)'

In [70]:
coffee_id=business_city_coffee["business_id"].tolist()
coffee_id

['lK-wuiq8b1TuU7bfbQZgsg',
 '8Hvp1tYKiQbBgGIwkCRK5g',
 'GvCblda5MdYBqNFBUlHuyg',
 'CfwrsG76Wm4iLS22v_wAcg',
 'PJ4eIqktcOD8W-uFBnPv-g',
 'rljaZWpdhJfrWAPT6kYb7g',
 'JOtv0NcBkvd9RO_D9cM3Qg',
 'v3kXkK9ga8jfgetr0Hb5Eg',
 'cUOX22xozKhnZM9-C0XIfA',
 '85Vov9WrC1yNkMeurAoRgQ',
 'uJIIgBmcIn-Ye-anrqK0DQ',
 '7YLvFqyFkIZkj5l6tnbqrw',
 '9Y7ymt0Kh0TkAo3F0LLN1Q',
 'u1qzJu0JwbKU-HjIFjq8kw',
 'cPZ5ugSrVrqnqD5ZsJhnpA',
 'Aes4MuZLnFqglFJNMZxRbw',
 '3gnQUW3pzKE32RenMfuf9A',
 'Mj5dXTDd6AndK8b4UVTNbw',
 'Zcz-ilAvtiZ36goz5Sa_Ow',
 'hq5MrQ8r3aBbaiGxfOyjsg',
 '0I8kPMnD_EWzGsDU9TGPNg',
 'v1G_bioP6POazUyQwHcz7Q',
 'QIXz6e0SZ4okLBFBsO79_Q',
 'RvpSsArnUXRZgC4-I7AyTQ',
 '-syU1HD-Iw_GEH8P5l03_A',
 '0u4Lmqi2hPCwwxJMbA7-ng',
 '-8F04F54iDT6VgWPCgybug',
 '-8PU0Z4Q2Nc-sY0eBj4DiQ',
 'pvoJj_qlDAS-mcHj-SUWTw',
 'DUYO3j2zAjsQ94HYqbFsCA',
 'n3AHTASqE5QCfuq0AQpT5Q',
 '6pG7n8Rx_7ZXeQQk690Nsw',
 'mg-6bfmZdM2AcdD8fXQkZA',
 'iIe6PLxsZDVyrKf4ziKxyw',
 '995cDn3wIwi4pw4N_uLSqw',
 '2VUhmFkIvARpEP9B5dZo9w',
 'nn8CMP4dDO0M2sA4jo309w',
 

In [71]:
review_city_coffee=review_city[review_city["business_id"].isin(coffee_id)]

In [72]:
review_city_coffee

Unnamed: 0               review_id                 user_id  \
31              31  sLbCQgT1ZJWB0e-XOVyS2g  I1mLZm-YSIarwaj0q9Njuw   
45              45  Kg5ncegiJ3utWRxDtDpi8g  1Dul59QEe-Q-7OQHTLOptw   
75              75  SkCsX9eyj-r4Hdw86dBOgQ  qWTeFVdxXS-H5l41XA_Lxw   
84              84  ezBhqKcg_4H84ii9Q7dGRw  i5WMqdq2vzi-U4W2aHhlNg   
94              94  YHYU3azWPb4UrKElyPO5QA  zS-MLrePVjxlkjv6RhtdYg   
...            ...                     ...                     ...   
941959      941959  5AiFkLlantJCPmposQp0hw  YB4_08Hn6l65kOeqdVhWpA   
941963      941963  PJaQsm03aJVWH9xg03WN9g  Kb0rYUpYzjJqoM7cAWZBAQ   
941977      941977  0-b3vZvU1-pld8jGHbCSEQ  R5P1nG4z0Q0OeVW3mhIKeA   
941978      941978  A_m_V1w_-YPchct8Ut5kWA  Ic6-gs1_FjrWGx6JIr95Mw   
942022      942022  YWfpUgsEKYg_J_dx3O-RPQ  XYCyePb6v-RKdm_THQQQbQ   

                   business_id  stars  useful  funny  cool  \
31      RvpSsArnUXRZgC4-I7AyTQ      3       0      0     0   
45      -8F04F54iDT6VgWPCgybug      4       1      0     0   
75      dpJbQLWxCgbNd9sboaed1w      5       2      0     0   
84      KsYAGLP4yC7FbfixaeCHsA      4       0      0     0   
94      Fozo0B-y42EhRMomR0K5vQ      5       0      0     0   
...                        ...    ...     ...    ...   ...   
941959  1Z6YrdtjD6hNzZRYpuNEsA      5       1      0     1   
941963  k0SB_IQ41iHessjK4hM2LA      1       0      0     0   
941977  MkLlb2944GtK07wy9u_wbA      1       0      0     0   
941978  T6qjr_kZSH8GECSoMbx8sQ      3      13      9    10   
942022  hWzbOi1-eS8PmPyZRV8Lew      5       0      0     0   

                                                                                                                                                                                                           text  \
31      It is a cute little place with very friendly staff; however, I went there on three different occasions and was disappointed with the quality of the food. I tried the cheddar and broccoli quiche wh...   
45      Super cute place! It's quite spacious with an open feel so doesn't feel like some crammed coffee shops.  The chicken sandwich was really tasty! The bubble tea I got was not my favorite, but not ba...   
75      THE UP-SIDE: Mojo's is wonderful, a call back to the days when coffee shops served great coffee and food, and didn't make you feel like a number. \n\nThis place feels like home. You can hang out a...   
84             Food is good, nothing extraordinary, but tasty and this restaurant wins all its points because of convenience, since it's open, cooking fresh food, 24/7. A favorite for late night, absolutely.   
94      The waffles here are to die for.  Having been to Belgium, I can say that these are almost just as delicious.  It can be a bit pricey but for all the waffle and toppings that you can get with it, i...   
...                                                                                                                                                                                                         ...   
941959  Cool little coffee shop located in the middle of downtown Cleveland. The staff is very friendly, helpful and talented with their latte art. There are not many seating options since the place is ve...   
941963  Bakery is super overpriced. smaller items are around 4-5 dollars, which is okay, but if you want a slightly bigger piece of bread it will cost you 8 dollars. Don't even bother with the drinks. Mos...   
941977  We went here last night in the drive thru and was told "We don't have nothing" and that was it. It was 9:40 pm\nThis store has been a nightmare for the last 8 years we've been here. It's the close...   
941978  Maybe I just do Coffee Tree Roasters wrong, but I have only had lackluster drinks with this chain.  I ordered a hot chocolate this week that was truly dreadful, and I have no idea what was in it t...   
942022                                                          AMAZING lattes and scones.  Nice rela

There are about 67000 reviews about coffee business. 